### DISCUSSION TITLE
Integration with Angular application

### DISCUSSION THREAD

**Discussion by Keshavabs** [molstar/molstar#1244] [Q&A] [answered]
Hi All,
We need your help with the integration of molstar npm package with our angular application.
We are trying to integrate it to view the molecular structures on our application. 

We integrated the npm package: https://www.npmjs.com/package/molstar

As a small sandbox solution, we created a new angular application and created a component that runs at path -> /view-structure/:token

This component has a route variable that can have two possible values -> ['pdb', 'xyz']

when route is /view-structure/pdb?structure=<_pdbId_>, it loads the structure from PDB.
when route is /view-structure/xyz?structure=<_file_name_of_file_stored_on_our_server_>, it loads the structure from our server.

view-structure.component.html
```<div id="view-structure"></div>```

view-structure.component.ts
`
path:string=  'https://server1.xyz.in/xyz/';
loadstructure(type, structure)
Viewer.create('view-structure', {
        layoutIsExpanded: true,
        layoutShowControls: false, //right side panel
        layoutShowRemoteState: false,
        layoutShowSequence: true,
        layoutShowLog: true,
        layoutShowLeftPanel: true,
        viewportShowExpand: false,
        viewportShowSelectionMode: true,
        viewportShowAnimation: false,
      }).then((viewer:any)=>{
          if(type == 'pdb') {
            viewer.loadPdb(structure);
          }
          else {
            viewer.loadStructureFromUrl(this.path + structure);
});
});`

We are trying out files with different formats hosted on our server, For some reason, _.cif_ files are loading and _.bcif_ files are not working.

We are able to download one more format: _.bcif.gz_ from pdb portal, but it is also not working, upon hosting on our server. It works fine when we try the open files option on the viewer.

`Error: Error: Unexpected token. Expected data_, loop_, or data name.`
Can some one please point out what is it that we are doing wrong?

**Comment by dsehnal** [comment, marked as answer]
`bcif` files require `isBinary` flag to be passed to `loadStructureFromUrl` (`loadStructureFromUrl(url: string, format: BuiltInTrajectoryFormat = 'mmcif', isBinary = false, options?: LoadStructureOptions & { label?: string })`)

**Reply by Keshavabs** [reply]
Thank you @dsehnal for quick response. That solves my original problem.

I have two more questions related to the file extensions, that I want to be answered.

 1. what would be the recommended file format to store the molecular data. is it mmcif ? if yes, how does the rest of the formats (mol2, cifcore etc) listed in the documentation, differ from mmcif ?  is it like cif and bcif formats come under the umbrella mmcif ?
 
2. if we can download the structure files from PDB in any format like .bcif.gz etc, and upload it on the viewer(using open files) and it works, then why does the same file not work when we host it on our server and do the viewer.loadStructureFromURL('...xyz.bcif.gz', true, {}); what is the reason for this ?

**Reply by dsehnal** [reply]
1. CIF and BCIF is in similar relationship as JSON and msgpack for example. Same data, different encoding. CIF is probably the most generic format currently and would be preferred for protein files. For other format considerations it really depends on your use case.
2. Opening the file checks the extension for GZ and if it's present automatically decompresses it. This is not done for URL and you have to tell the viewer explicitly that the data is compressed similar to how you had to specify `isBinary` flag. Other option is to include `content-encoding: gzip` (hope I got it right from the top of my head) to response headers from your server and the browser will decompress the data for you automatically.

**Reply by Keshavabs** [reply]
Thanks for answering my queries 👍
